File: crates/openpit/src/param/side.rs
// Copyright The Pit Project Owners. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please see https://github.com/openpitkit and the OWNERS file for details.

/// Represents the side of a trade or order.
///
/// This enum is `Copy` and intended to behave like a small value type.
/// It does not encode numeric meaning implicitly — use methods like
/// [`Side::sign`] if a signed representation is needed.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Side {
    /// Buy side (long direction).
    Buy,

    /// Sell side (short direction).
    Sell,
}

impl Side {
    /// Returns `true` if the side is [`Side::Buy`].
    #[inline]
    pub fn is_buy(self) -> bool {
        match self {
            Side::Buy => true,
            Side::Sell => false,
        }
    }

    /// Returns `true` if the side is [`Side::Sell`].
    #[inline]
    pub fn is_sell(self) -> bool {
        match self {
            Side::Buy => false,
            Side::Sell => true,
        }
    }

    /// Returns the opposite trading side.
    ///
    /// # Examples
    ///
    /// ```
    /// use openpit::param::Side;
    ///
    /// assert_eq!(Side::Buy.opposite(), Side::Sell);
    /// assert_eq!(Side::Sell.opposite(), Side::Buy);
    /// ```
    #[inline]
    pub fn opposite(self) -> Self {
        match self {
            Side::Buy => Side::Sell,
            Side::Sell => Side::Buy,
        }
    }

    /// Returns a signed representation of the side.
    ///
    /// `+1` for [`Side::Buy`], `-1` for [`Side::Sell`].
    ///
    /// Intended for financial calculations (e.g., signed quantity or exposure).
    #[inline]
    pub fn sign(self) -> i8 {
        match self {
            Side::Buy => 1,
            Side::Sell => -1,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Side;

    #[test]
    fn side_predicates_work() {
        assert!(Side::Buy.is_buy());
        assert!(!Side::Buy.is_sell());
        assert!(Side::Sell.is_sell());
        assert!(!Side::Sell.is_buy());
    }

    #[test]
    fn opposite_returns_other_side() {
        assert_eq!(Side::Buy.opposite(), Side::Sell);
        assert_eq!(Side::Sell.opposite(), Side::Buy);
    }

    #[test]
    fn sign_matches_direction() {
        assert_eq!(Side::Buy.sign(), 1);
        assert_eq!(Side::Sell.sign(), -1);
    }
}

File: bindings/go/param/side.go
// Copyright The Pit Project Owners. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please see https://github.com/openpitkit and the OWNERS file for details.

package param

import (
	"fmt"

	"go.openpit.dev/openpit/internal/native"
	"go.openpit.dev/openpit/pkg/optional"
)

// Side represents the side of a trade or order.
type Side native.ParamSide

const (
	// SideBuy means buy side.
	SideBuy Side = native.ParamSideBuy
	// SideSell means sell side.
	SideSell Side = native.ParamSideSell
)

// NewSideFromHandle creates an optional Side from a native handle.
func NewSideFromHandle(v native.ParamSide) optional.Option[Side] {
	switch v {
	case native.ParamSideSell:
		return optional.Some(SideSell)
	case native.ParamSideBuy:
		return optional.Some(SideBuy)
	case native.ParamSideNotSet:
		return optional.None[Side]()
	default:
		panic(fmt.Sprintf("unknown native ParamSide value %d", v))
	}
}

// IsBuy returns true when side is buy.
func (v Side) IsBuy() bool {
	return v == SideBuy
}

// IsSell returns true when side is sell.
func (v Side) IsSell() bool {
	return v == SideSell
}

// Opposite returns the opposite side.
func (v Side) Opposite() Side {
	if v == SideBuy {
		return SideSell
	}
	return SideBuy
}

// Sign returns +1 for buy and -1 for sell.
func (v Side) Sign() int8 {
	if v == SideBuy {
		return 1
	}
	return -1
}

// String returns a human-readable representation of the side.
func (v Side) String() string {
	if v == SideBuy {
		return "buy"
	}
	return "sell"
}

// Handle returns the underlying native handle.
func (v Side) Handle() native.ParamSide {
	return native.ParamSide(v)
}
